### DISCUSSION TITLE
Backup/Secondary MX

### DISCUSSION THREAD

**Discussion by FZnachor** [emersion/maddy#653] [Q&A] [answered]
Hello, I would like to set up a second maddy server as a backup for the main maddy server as I had it with postfix.

It worked in such a way that if the main server is unavailable, the second server stores the message and periodically tries to send it to the main server.

Is this possible using maddy? Thanks.

**Comment by foxcpp** [comment, marked as answer]
`target.queue` can be used instead of a local storage. The only problem is making sure secondary MX knows what emails are valid, it would be unwise to just accept any and all emails independently of the account name. Also you need to setup a secondary endpoint with authentication on primary MX to accept relayed messages while bypassing security checks (SPF, for example, would almost definitely fail and cause false positives).

For example, on secondary MX:
```
target.queue primary_mx_queue {
    target lmtp tls://mx1.example.org:2525 { # https://maddy.email/reference/targets/smtp/
        require_tls # You probably want this.
        auth plain "USERNAME" "PASSWORD" # Define authentication data that would be accepted by LMTP endpoint, see below.
    }

    autogenerated_msg_domain $(primary_domain)
    bounce {
        destination postmaster $(local_domains) {
            deliver_to &local_routing
        }
        default_destination {
            reject 550 5.0.0 "Refusing to send DSNs to non-local addresses"
        }
    }
}

msgpipeline local_routing {
    destination_in ???? { # would need to define a check here somehow
        deliver_to &primary_mx_queue
    }

    default_destination {
        reject 550 5.1.1 "User doesn't exist"
    }
}

smtp tcp://0.0.0.0:25 {
    limits {
        # Up to 20 msgs/sec across max. 10 SMTP connections.
        all rate 20 1s
        all concurrency 10
    }

    # Must run the same security checks as the primary MX.
    dmarc yes
    check {
        require_mx_record
        dkim
        spf
    }

    source $(local_domains) {
        reject 501 5.1.8 "Use Submission for outgoing SMTP"
    }
    default_source {
        deliver_to &local_routing
    }
}
```

On primary MX:
```
endpoint.lmtp tls://0.0.0.0:2525 {
  auth pass_table static {
    entry "USERNAME" "HASHED_PASWORD" # hashed using "maddy hash" command
  } 
  deliver_to &local_routing
}
```

**Comment by dehnhard** [comment]
Thanks to the advice above I successfully setup a secondary MX server.

Both primary and secondary server have rspamd configured. Spam received and detected on the secondary MX is marked correctly (X-Spam-Flag/Score), but not quarantined (moved to the junk folder) when delivered to the primary mail server.

Is there a way to achieve this?
